packages/public/server/src/module/Alias/src/Controller/AliasController.php

```diff
@@ -22,71 +22,149 @@
-            );
-        } catch (AliasNotFoundException $e) {
-            $this->getResponse()->setStatusCode(404);
-            return false;
+            $object = $this->uuidManager->getUuid($id, true);
+        } catch (NotFoundException $e) {
+            return $this->notFoundResponse();
+        }
+        $normalized = $this->normalizer->normalize($object);
+        $url = $this->getUrlOfNormalizedObject($normalized);
+        $response = $this->routerResponse($url);
+
+        if (!$this->getRequest()->isXmlHttpRequest()) {
+            return $response;
         }
 
-        $router = $this->getServiceLocator()->get('Router');
+        // Note: The behavior below is needed for injections. Basically, we override the response when we have an XHR and return JSON instead.
+
+        if ($response instanceof JsonModel) {
+            $response = new ViewModel(['data' => $response->getVariables()]);
+            $response->setTemplate('normalizer/json');
+        }
+
+        $view = new ViewModel([
+            'id' => $object->getId(),
+            'type' => $normalized->getType(),
+            'url' => $url,
+            '__disableTemplateDebugger' => true,
+        ]);
+
+        $view->addChild($response, 'response');
+        $view->setTemplate('normalizer/ref');
+        $view->setTerminal(true);
+        return $view;
+    }
+
+    private function getUrlOfNormalizedObject($normalized)
+    {
+        $routeName = $normalized->getRouteName();
+        $routeParams = $normalized->getRouteParams();
+        /** @var Url $urlPlugin */
+        $urlPlugin = $this->url();
+        return $urlPlugin->fromRoute(
+            $routeName,
+            $routeParams,
+            null,
+            null,
+            false
+        );
+    }
+
+    private function routerResponse($url)
+    {
         $request = new Request();
         $request->setMethod(Request::METHOD_GET);
-        $request->setUri($source);
-        $routeMatch = $router->match($request);
+        $request->setUri($url);
+        $routeMatch = $this->router->match($request);
 
         if ($routeMatch === null) {
-            $this->getResponse()->setStatusCode(404);
-            return false;
+            return $this->notFoundResponse();
         }
 
         $this->getEvent()->setRouteMatch($routeMatch);
-        $params = $routeMatch->getParams();
+        $params = array_merge($routeMatch->getParams(), [
+            'forwarded' => true,
+            'isXmlHttpRequest' => $this->getRequest()->isXmlHttpRequest(),
+        ]);
         $controller = $params['controller'];
-        $return = $this->forward()->dispatch(
-            $controller,
-            ArrayUtils::merge($params, [
-                'forwarded' => true,
-            ])
-        );
+        return $this->forward()->dispatch($controller, $params);
+    }
 
-        return $return;
+    private function notFoundResponse()
+    {
+        $this->getResponse()->setStatusCode(404);
+        return false;
```

Comment:
Follow up: I would move this function to an helper Mixin / base class so that we can use it in other controllers as well.

Reply:
We can do that when we need it somewhere else, too.